### PR TITLE
MDEV-24670 memory pressure - eventfd rather than pipe

### PR DESCRIPTION
**Pull request by grooverdan**
After discovering eventfds for #2807 it simplifies the current implementation a little here too.

Like #2807 we don't get to read from the eventfd in the quit case as the m_abort flag is set.

- [x] *The Jira issue number for this PR is: MDEV-24670*

## Description

Eventfds have a simpler interface and are one file descriptor rather than two.

Reuse the patten of the accepting socket connections by testing for abort after a poll returns. This way the same event descriptor can be used for Quit
and debugging trigger.

Also correct the registration of mem pressure file descriptors.

## How can this PR be tested?

mem_pressure test already.

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch.*
- [X] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

## PR quality check
- [X] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/11.0/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [X] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

### CODE REVIEW

File: storage/innobase/buf/buf0buf.cc [hunk 1]
```diff
@@ -752,14 +753,16 @@ based off https://www.kernel.org/doc/html/latest/accounting/psi.html#pressure-in
...
   struct pollfd m_fds[3];
   nfds_t m_num_fds;
-  int m_pipe[2]= {-1, -1};
+  int m_event_fd= -1;
 
   std::thread m_thd;
-  /* mem pressure garbarge collection resticted to interval */
+  /* mem pressure garbage collection restricted to interval */
   static constexpr ulonglong max_interval_us= 60*1000000;
+  static bool m_abort;
```
Comment (dr-m) [inline storage/innobase/buf/buf0buf.cc:764]: If this were declared non-`static` we would not need a separate initialization. Maybe, to silence any ThreadSanitizer warning, this should be `Atomic_relaxed<bool>`.

File: storage/innobase/buf/buf0buf.cc [hunk 2]
```diff
@@ -921,23 +923,17 @@ void mem_pressure::pressure_routine(mem_pressure *m)
...
-            goto shutdown;
-#ifdef UNIV_DEBUG
-          case 'G':
-            buf_pool.garbage_collect();
-#endif
-          }
+        uint32 u;
+        /* we haven't aborted, so this must be a debug trigger */
+        if (read(p.fd, &u, sizeof(u)) >=0)
+          buf_pool.garbage_collect();
```
Comment (dr-m) [inline storage/innobase/buf/buf0buf.cc:932]: We wrote `uint64_t` but are reading only `uint32`. Could this be a single-byte read and write?

**Review by dr-m — COMMENTED** [review]
Looks OK with a couple cleanups.

File: storage/innobase/buf/buf0buf.cc (hunk 2 above)
Comment (grooverdan) [reply storage/innobase/buf/buf0buf.cc]: eventfd prescribes 64 bit

File: storage/innobase/buf/buf0buf.cc (hunk 1 above)
Comment (grooverdan) [reply storage/innobase/buf/buf0buf.cc]: missed the obvious non-static implementation, thanks.

File: storage/innobase/buf/buf0buf.cc [hunk 3]
```diff
@@ -921,23 +922,17 @@ void mem_pressure::pressure_routine(mem_pressure *m)
...
-          case 'Q':
-            goto shutdown;
-#ifdef UNIV_DEBUG
-          case 'G':
-            buf_pool.garbage_collect();
-#endif
-          }
+        uint64_t u;
+        /* we haven't aborted, so this must be a debug trigger */
+        if (read(p.fd, &u, sizeof(u)) >=0)
```
Comment (dr-m) [inline storage/innobase/buf/buf0buf.cc:930]: We had written 8 bytes, but here (depending on the kernel-side implementation) we could theoretically read that in multiple requests, so a single debug event could theoretically trigger multiple garbage collections. Not a big deal.

**Review by dr-m — APPROVED** [review]

File: storage/innobase/buf/buf0buf.cc (hunk 3 above)
Comment (grooverdan) [reply storage/innobase/buf/buf0buf.cc:930]: the way its describes seems rather strict on 8 bytes. But who doesn't like a bit of debug fuzzing :-)

**Pull request merged**